querymsc.py

```diff
@@ -505,5 +507,124 @@ def query_rain_rate_one_day_1_50(self, session):
                  )
- 
-        return query
\ No newline at end of file
+  
+        return query
+
+class Gumbel:
+    """This class is meant to do all the fitting
+    of Gumbel distributions to extract design values.
+    
+    Attributes:
+        return_period (int): return period in years
+        min_fit (int): minimum number of years to consider
+            fitting Gumbel.
+    """
+
+    def __init__(self, return_period, min_fit=10):
+
+        self.return_period = return_period
+        self.min_fit = min_fit
+
+    def fit(self, x):
+        """Function to get L-moments to estimate the parameters
+        of a gumbel distribution. This method uses the 
+        right-skewed Gumbel distribution. Method mirrors 
+        Hosking 1990.
+        -----------------------------------
+        Args:
+            x (pandas Series): Series containing the annual
+                grouped extreme values for a given 
+                station over a range of years.
+        Returns:
+            (xi, alpha) (tuple): estimated parameters of gumbel
+                distribution if N_min criteria met
+            NaN (numpy NaN object): if N_min criteria is not met 
+        """
+
+        N = x.shape[0]
+
+        # euler-mascheroni constant
+        euler = 0.5772156649
+
+        if N >= self.min_fit:
+            # create fitted gumbel object to
+            # distribution of extreme values
+            paras = distr.gum.lmom_fit(x)
+
+            # extract L-moments
+            lmoments = distr.gum.lmom(nmom=2, **paras)
+
+            # calculate estimators of gumbel
+            # parameters
+            alpha = lmoments[1]/np.log(2)
+            xi = lmoments[0] - euler*alpha 
+
+            return xi, alpha
+        else:
+            return np.nan, np.nan
+
+    def get_design_value(self, xi, alpha):
+        """Get the design value from explicit expression
+        derived from a right-skewed Gumbel distribution.
+        See methods.pdf for more details. Separate from
+        Gumbel fitting in case user defines custom
+        estimators. 
+        -----------------------------------
+        Args:
+            xi (float): Estimated value of the location parameter
+                (first l-moment)
+            alpha (float): Estimated value of the scale parameter
+                (second l-moment).
+        Returns:
+            design_val (float): design value using Gumbel parameters
+        """
+
+        # return frequency, i.e. inverse of return period 
+        f_r = 1.0/self.return_period
+
+        # simplify long expression
+        simp = (1.0-f_r) + np.exp(-np.exp((xi/alpha)))
+
+        # final expression for design value
+        design_val = xi - alpha*np.log(-np.log(simp))
+
+        return design_val
+
+    def get_fit_transform(self, x):
+        """Extract the design value.
+        -----------------------------
+        Args:  
+            x (pandas Series): Series containing the annual
```

Comment:
For any place where your input is a `pandas Series` or `DataFrame`, it would be helpful to have a bit more detail about what data should be contained in that. Those data structures are highly flexible, which is great for exploration, but the fact that they're flexible means that there is a wide range of valid objects that might not work in this function.

Is `x` just an array of numbers? What value does the `pandas Series` add in this case? It's often best to use the simplest data structure that accomplishes the task. Your tests will be simpler, it will be easier to use the code, etc.

Reply:
Yes - it is a 
> Numerical series of annual extremes grouped by station.

While `fit`, and `get_fit_transform` can handle either a `ndarray` or `pandas Series`, `fit_transform` groups `DataFrame` with a whole bunch of observations for a bunch of stations into a format where the fitting can take place quickly on _groups within_ the `DataFrame`.

In particular, this line of code
```
df_new = df.join((df.groupby(station_variable)[variable]
                            .apply(self.get_fit_transform)), 
                         on=station_variable,
                         rsuffix='_dval')
```

Is an essential preparation bit to perform the gumbel fitting on a `DataFrame` that contains annual extremes for a bunch of stations over a bunch of different timescales. It operates in chunks of stations, then assigns the return value of  `get_fit_transform` to a row in the original `DataFrame`.

It's kind of a complicated concept that is hard to represent in the documentation, but I've reworded it a bit to try and clarify.

Reply:
Perhaps
> Numerical series of annual extremes of a unique station

is more clear?